### DISCUSSION TITLE
Custom interest coordinates for smartcrop

### DISCUSSION THREAD

**Discussion by ldbglobe** [libvips/libvips#3121] [Q&A]
I often deal with photography with lot of detail on all the picture. And the smartcrop interest like entropy or interest fail to found a real center of interest resulting of a centered crop of the original file.

In my previous image resizer implementation (eating too much CPU and memory for my current usage) I let the user indicate the location of the main COI (center of interest) along X and Y axis (a value between 0-100%) and use it as parameters in the cropping process.

Or if no custom COI was set a smart content detector was used trying to find something.

It will be interesting to propose a similar possibility in a futur version. Allowing to use an X%,Y% combination as "interest" parameter

Best regards
Mathieu

**Comment by jcupitt** [comment]
Hi Mathieu,

Sure, let's tag this as an enhancement.

**Comment by jcupitt** [comment]
Actually, thinking about it, you can implement this efficiently already using python/ruby/php/etc., eg. (untested):

```python
import pyvips

# load and resize the longest (edit: this used to say smallest, thanks kip) axis to 100 pixels
thumb = pyvips.Image.thumbnail("huge.jpg", 100, crop="all")

# now crop to 100x100 using a supplied percent
if thumb.width > 100:
    thumb = thumb.crop((thumb.width - 100) * (x_percent / 100), 0, 100, 100)
else:
    thumb = thumb.crop(0, (thumb.height - 100) * (y_percent / 100), 100, 100)
```

Because libvips is demand-driven, thumbnail will only compute the pixels you end up using in the crop, so there would be no speed or memory saving if we added the option to thumbnail directly.

Would that fix your problem?

**Comment by ldbglobe** [comment]
Your solution work like a charm
And I now understand better the vips philosophy 👍

Thanks you very much

**Comment by kipcole9** [comment]
> load and resize the smallest axis to 100 pixels

That's interesting!  The docs indicate this parameter is `width` and hence I assumed it resized to that width (keeping the aspect ratio by default).  Does it truly mean "resize the smallest axis" instead?

My experiment seems to suggest it actually resizes the longest axis to the specified size?
```elixir
# Open the image and get its shape in the form {width, height, bands}
iex> {:ok, i} = Image.open "./test/support/images/Singapore-2016-09-5887.jpg"
iex> Image.shape(i)
{1000, 590, 3}
# Thumbnail it with the "width" of 100 and get its shape
# Its resized the longest axis (width) to be 100 and retained aspect ration
iex> Image.thumbnail!(i, 100) |> Image.shape()
{100, 59, 3}
# Now rotate the image 90 degrees and get its shape
iex> {:ok, rotated} = Image.rotate(i, 90)
iex> Image.shape(rotated)
{590, 1000, 3}
# Resize the rotated image. It resizes the longest axis
# to 100px and keeps the aspect ration
iex> Image.thumbnail!(rotated, 100) |> Image.shape
{59, 100, 3}
```

**Reply by jcupitt** [reply]
Oop, it's longest, duh, you're right. I'll edit my answer to reduce confusion.
